### PR TITLE
move: create Arc app state for source service

### PR DESCRIPTION
**Pull request by rvantonder**
## Description 

The source service has two long running threads:

- the server thread that serves source when requests come in
- a thread that monitors on-chain transactions for upgrades (`fn watch_for_upgrades`)

The second thread needs access to the server thread state, in order to invalidate (and later update) source code when it observes an on-chain upgrade. This PR simply does the plumbing to share the server state as a mutable reference in `watch_for_upgrades`. The state is not used yet, see follow up PRs for more.

## Test Plan 

N/A no functional changes

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-source-validation-service/src/lib.rs [hunk 1]
```diff
@@ -361,7 +361,10 @@ pub async fn verify_packages(config: &Config, dir: &Path) -> anyhow::Result<Netw
     Ok(lookup)
 }
 
-pub async fn watch_for_upgrades(config: &Config) -> anyhow::Result<()> {
+pub async fn watch_for_upgrades(
+    config: &Config,
+    _app_state: Arc<RwLock<AppState>>,
+) -> anyhow::Result<()> {
```
Comment (rvantonder) [inline crates/sui-source-validation-service/src/lib.rs:367]: Main objective: `watch_for_upgrades` has a handle to `_app_state`. It is unused right now, see follow up PRs for more.

File: crates/sui-source-validation-service/src/lib.rs [hunk 2]
```diff
@@ -394,16 +397,19 @@ pub struct AppState {
...
+        .route("/api", get(api_route))
         .layer(
             ServiceBuilder::new().layer(
                 tower_http::cors::CorsLayer::new()
                     .allow_methods([Method::GET])
                     .allow_origin(tower_http::cors::Any),
             ),
-        );
+        )
+        .with_state(app_state);
```
Comment (rvantonder) [inline crates/sui-source-validation-service/src/lib.rs:412]: Just changes in the server to account for the new app state being `RwLock`-able

File: crates/sui-source-validation-service/src/main.rs [hunk 3]
```diff
@@ -25,7 +28,8 @@ pub async fn main() -> anyhow::Result<()> {
     let sources = initialize(&package_config, tmp_dir.path()).await?;
     info!("verification complete in {:?}", start.elapsed());
     info!("serving on {}", host_port());
-    serve(AppState { sources })?
-        .await
-        .map_err(anyhow::Error::from)
+    let app_state = Arc::new(RwLock::new(AppState { sources }));
+    let app_state_copy = app_state.clone();
+    tokio::spawn(async move { watch_for_upgrades(&package_config, app_state).await });
+    serve(app_state_copy)?.await.map_err(anyhow::Error::from)
```
Comment (rvantonder) [inline crates/sui-source-validation-service/src/main.rs:34]: Changes to share state ref across both `watch_for_upgrades` and `serve` threads

**Review by tnowacki — APPROVED** [review]
I don't really know what is going on, but it seems reasonable to me!

**Review by stefan-mysten — APPROVED** [review]

**Comment by rvantonder** [comment]
> I don't really know what is going on, but it seems reasonable to me!

Yep, all good, ✅ is all I need here. Happy to walk through things but I don't think these changing are worth paging in for others atm.

**Pull request merged**